partiql-lang/src/main/kotlin/org/partiql/lang/eval/physical/PhysicalPlanCompilerAsyncImpl.kt

```diff
@@ -0,0 +1,1908 @@
+            is PartiqlPhysical.Expr.Ne -> compileNe(expr, metas)
+            is PartiqlPhysical.Expr.Or -> compileOr(expr, metas)
+
+            // unary
+            is PartiqlPhysical.Expr.Not -> compileNot(expr, metas)
+            is PartiqlPhysical.Expr.Pos -> compilePos(expr, metas)
+            is PartiqlPhysical.Expr.Neg -> compileNeg(expr, metas)
+
+            // other operators
+            is PartiqlPhysical.Expr.Concat -> compileConcat(expr, metas)
+            is PartiqlPhysical.Expr.Call -> compileCall(expr, metas)
+            is PartiqlPhysical.Expr.NullIf -> compileNullIf(expr, metas)
+            is PartiqlPhysical.Expr.Coalesce -> compileCoalesce(expr, metas)
+
+            // "typed" operators (RHS is a data type and not an expression)
+            is PartiqlPhysical.Expr.Cast -> compileCast(expr, metas)
+            is PartiqlPhysical.Expr.IsType -> compileIs(expr, metas)
+            is PartiqlPhysical.Expr.CanCast -> compileCanCast(expr, metas)
+            is PartiqlPhysical.Expr.CanLosslessCast -> compileCanLosslessCast(expr, metas)
+
+            // sequence constructors
+            is PartiqlPhysical.Expr.List -> compileSeq(ExprValueType.LIST, expr.values, metas)
+            is PartiqlPhysical.Expr.Sexp -> compileSeq(ExprValueType.SEXP, expr.values, metas)
+            is PartiqlPhysical.Expr.Bag -> compileSeq(ExprValueType.BAG, expr.values, metas)
+
+            // bag operators
+            is PartiqlPhysical.Expr.BagOp -> compileBagOp(expr, metas)
+            is PartiqlPhysical.Expr.BindingsToValues -> compileBindingsToValues(expr)
+            is PartiqlPhysical.Expr.Pivot -> compilePivot(expr, metas)
+            is PartiqlPhysical.Expr.GraphMatch -> TODO("Physical compilation of GraphMatch expression")
+            is PartiqlPhysical.Expr.Timestamp -> TODO()
+        }
+    }
+
+    private suspend fun compileBindingsToValues(expr: PartiqlPhysical.Expr.BindingsToValues): PhysicalPlanThunkAsync {
+        val mapThunk = compileAstExpr(expr.exp)
+        val bexprThunk: RelationThunkEnvAsync = bexperConverter.convert(expr.query)
+
+        val relationType = when (expr.metas.containsKey(IsOrderedMeta.tag)) {
+            true -> RelationType.LIST
+            false -> RelationType.BAG
+        }
+
+        return thunkFactory.thunkEnvAsync(expr.metas) { env ->
+            // we create a snapshot for currentRegister to use during the evaluation
+            // this is to avoid issue when iterator planner result
+            val currentRegister = env.registers.clone()
+            val elements: Flow<ExprValue> = flow {
+                env.load(currentRegister)
+                val relItr = bexprThunk(env)
+                while (relItr.nextRow()) {
+                    emit(mapThunk(env))
+                }
+            }
+            when (relationType) {
+                RelationType.LIST -> ExprValue.newList(elements.toList())
+                RelationType.BAG -> ExprValue.newBag(elements.toList())
```

Comment:
Discussed offline and including a summary of the discussion. To summarize, making the async evaluator lazy would necessitate a larger refactor of the evaluator. We will keep the async evaluator as it is and revisit if we find we need the previous, lazy materialization behavior.

For more context, we use `Flow`s in this block (as well as other parts of the async evaluator) rather than `Sequence`s because code within a `Flow` block can call suspend functions. `Flow`s act somewhat differently than sequences in that to get the computed value back from a `Flow`, you need to "collect" the flow using a function like [`Flow.collect`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/collect.html), which appears to eagerly get all items. From the docs on [`Flow`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-flow/):
> Terminal operators on the flow are either suspending functions such as [collect](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-flow/collect.html), [single](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/single.html), [reduce](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/reduce.html), [toList](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/index.html), etc. or [launchIn](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/launch-in.html) operator that starts collection of the flow in the given scope. They are applied to the upstream flow and trigger execution of all operations. Execution of the flow is also called collecting the flow and is always performed in a suspending manner without actual blocking. Terminal operators complete normally or exceptionally depending on successful or failed execution of all the flow operations in the upstream.

Making an `ExprValue` that accepts a `Flow` rather than a list/sequence will not work either. `ExprValue` implements `Iterable<ExprValue>`, which necessitates overriding `iterator()`. Since `iterator()` is synchronous and calls to `Flow.collect()` (or similar terminal operators) are async, this approach is not suitable.

One possible solution could be to make `PartiQLResult.Value`'s parameter `value` a `Flow<ExprValue>` rather than an `ExprValue`. This could ensure the materialization is left to the consuming application when they call `.collect` or `.toList()` rather than within the async evaluator. But this change would require a larger refactor and possible duplication of `PartiQLResult`.